torch/_dynamo/guards.py

```diff
@@ -468,11 +469,54 @@ def TENSOR_MATCH(self, guard: Guard):
                 for term in terms:
                     real_value = self.get(tensor_name + "." + term)
                     code.append(f"{tensor_name}.{term} == {real_value}")
-                self._produce_guard_code(guard, code)
             else:
                 self.tensor_check_names.append(tensor_name)
                 self.tensor_check_examples.append(value)
 
+            # A frame is valid for reuse with dynamic dimensions if the new dynamic dimensions are a
+            # strict subset of the old.
+            #
+            # The logic here is as follows:
+            #
+            # Every mark_dynamic directive is a user-knows-best command, which can incur a raise at tracing
+            # time if we find guards that run counter to the user directive.
+            #
+            # If the frame is compiled with any marked dynamic indices, let's call that set of indices X.
+            # When we evaluated inputs against the guards, given the same tensor with potentially new dynamic indices, let's call that set Y.
+            # Y must be a strict subset of X.
+            #
+            # This is the case because any newly introduced mark_dynamic directives have a chance of
+            # incurring raises, failing compilation. Any existing mark_dynamic indices that we lost are safe to lose
+            # as all it means is that we have gotten rid of a user directive which could incur a raise at compile time.
+            # In the case of when there is no Y, that is, there are no dynamic indices marked at all, the frame is safe to reuse
+            # as an empty set is a safe degeneration - that is, a strictly static tensor is always valid for a frame compiled with that same
+            # tensor + more onerous user directives.
```

Comment:
I understand the explanation here, but I think it can be explained more clearly. The problem with this comment is that it doesn't explain what the desirable system level invariant is. In this case, the safety invariant is "if compiling a frame with explicit dynamic dims X could cause an exception, we MUST NOT skip compiling." This then motivates the rest of the reasoning.

BTW, there's something else we should probably discuss in the API documentation, which is that if a user compiles a program once with an input tensor with dynamic dims, should they expect subsequent calls to not recompile, even if the new input tensors don't have dynamic dim annotations at all? I used to think the answer is yes, but on further consideration I think the answer must be no in our current implementation, because our previous compiled-for-dynamic code may fail guards for other reasons, and then if we recompile and only consult the dynamic dim annotation at time of compilation, of course we won't compile with dynamic dims if you didn't set the attributes correctly. But perhaps a different semantics could preserve this: e.g., if the dynamic dim setting was "sticky" for the frame and persisted some other mechanism. This isn't really something we do today, but when we implement dynamic by default by observing when a dim varies over time, this would logically roll into that (setting something as dynamic just says to immediately assume the dim varied over time. Future calls don't need the annotation.)

Reply:
>  but I think it can be explained more clearly. 

Agreed

> should they expect subsequent calls to not recompile, even if the new input tensors don't have dynamic dim annotations at all? 

I think so? 

> I think the answer must be no in our current implementation, because our previous compiled-for-dynamic code may fail guards for other reasons, and then if we recompile and only consult the dynamic dim annotation at time of compilation

If the frame failed compilation for other reasons, then on recompilation it will recheck the dynamism of the tensor, and 
store that in its guards. I guess I do not understand the motivating problem case here?

Reply:
Suppose you have

```
x = torch.randn(2,3,4)
y = torch.randn(3, dtype=torch.long)
torch._dynamo.mark_dynamic_dim(x, 0)
opt_f(x, y)
x = torch.randn(2,3,4)
y = torch.randn(3, dtype=torch.int)
opt_f(x, y)
```

The second opt_f clearly needs to recompile. But does the 0th dim of x get marked dynamic? Naively, I'd think no, but what if I had:

```
x = torch.randn(2,3,4)
y = torch.randn(3, dtype=torch.long)
opt_f(x, y)
x = torch.randn(3,3,4)
opt_f(x, y)
x = torch.randn(4,3,4)
opt_f(x, y)
x = torch.randn(4,3,4)
y = torch.randn(3, dtype=torch.int)
opt_f(x, y)
```

In the hypothetical universe where PT2 automatically realizes that the first dim is varying and decides to mark it as dynamic so we don't recompile, it seems bad for the last call (which fails guard for unrelated reasons) to not also be immediately marked as dynamic.

Reply:
But a guard is not "automatically realizes that the first dim is varying and decides to mark it as dynamic"? Right? There's not auto-flagging. Rather, what you end up doing is reusing a frame, *as if* you had it marked. This is an important distinction, because if an unrelated guard fails, you will recompile with the current state of mark_dynamic, not some old remembered state we automatically apply.

I do not understand why it seems bad... preserving the dynamism and trying to automatically apply it feels like it opens us up to a ton of edge cases - any recompile should just honor the values of the trace as they are at the time of recompile.

Reply:
To elaborate further - this is a niche feature for power users, ones who understand the implications of dynamic shapes and dynamism. 

Preserving mark_dynamic directives many callsites later, with different tensors, automatically, assumes too much. 

In this case, if we mark_dynamic, on a tensor... the pythonic and expected behavior is we only changed it for that object. 

Saving this state, to me, feels surprising. It also now means that, in a case where a user wants to specifically do the opposite, in the assume story, what API do they have for explicitly forcing a compile *without* a dynamic dim at `x` `0` in your example above? deleting a dynamic dim from a tensor won't work, it's not there... do they torch._dynamo.reset()? Do we give them a clear api? It just feels like too much. Versus, in today's world, you just have to remember to call mark_dynamic on that input if you want to preserve that same behavior - a much clearer and simpler API.

Reply:
What I am trying to do is preemptively harmonize this API (the power API, mark dynamic dim) with the automatic dynamic enablement API (switch to dynamic if you notice you're recompiling). I agree that the stateful API is more complex. But this is *precisely* what we are going to do with automatic dynamic enablement. If we think automatic dynamic enablement is a good idea (everyone seemed to think so when @jansel suggested it), then I think there's a strong incentive to keep mark dynamic dim consistent with it as well.

Reply:
I see. I am not 100% sold on this being a good state for mark_dynamic. I do think automatic enablement is a good thing. That being said, I am happy to disagree and commit here, you are historically right about these things, and tend to have a better vision of what we should offer. 

Let me know what you think we should do here.

Reply:
We discussed offline. We will land it as is now until the automatic dynamic enablement workstream is started, and then revisit.